linker/shared/src/test/scala/org/scalajs/linker/IncrementalTest.scala

```diff
@@ -187,79 +187,222 @@ class IncrementalTest {
+      ApplyStatic(EAF, FooClass, targetMeth, Nil)(NoType)
+
+    def classDefs(step: Int): List[ClassDef] = {
+      val stepDependentMembers = step match {
+        case 0 =>
+          List(
+            methDef(run1, consoleLog(str("a1"))),
+            methDef(run2, consoleLog(str("a2")))
+          )
+        case 1 =>
+          List(
+            methDef(run1, consoleLog(str("b1")))
+          )
+        case 2 =>
+          List(
+            methDef(run1, consoleLog(str("c1"))),
+            methDef(run2, consoleLog(str("c2")))
+          )
+      }
+
+      val stepDependentMainStats = step match {
+        case 0 => List(callMeth(run1), callMeth(run2))
+        case 1 => List(callMeth(run1))
+        case 2 => List(callMeth(run1), callMeth(run2))
+      }
+
+      List(
+        classDef(FooClass, superClass = Some(ObjectClass),
+            memberDefs = trivialCtor(FooClass) :: stepDependentMembers),
+
+        mainTestClassDef(Block(stepDependentMainStats))
+      )
+    }
+
+    def moduleInitializers(step: Int): List[ModuleInitializer] =
+      MainTestModuleInitializers
+
+    testIncrementalSteps("issue 4416 static variant", steps = 3, classDefs(_),
+        moduleInitializers(_))
   }
 
 }
 
 object IncrementalTest {
 
-  def testIncremental(
+  type Input = (Seq[ClassDef], List[ModuleInitializer])
+
+  def testIncrementalBidirectional(
       classDefs: Boolean => Seq[ClassDef],
       moduleInitializers: Boolean => List[ModuleInitializer],
       config: StandardConfig = StandardConfig())(
       implicit ec: ExecutionContext): Future[Unit] = {
+
+    def testOneDirection(contextMessage: String, forward: Boolean): Future[Unit] = {
+      testIncrementalSteps(contextMessage, steps = 2,
+          step => classDefs(step == 0), step => moduleInitializers(step == 0),
+          config)
+    }
+
     for {
-      _ <- testIncrementalStep(backward = false, classDefs, moduleInitializers, config)
-      _ <- testIncrementalStep(backward = true, classDefs, moduleInitializers, config)
+      _ <- testOneDirection("forward", forward = true)
+      _ <- testOneDirection("backward", forward = false)
     } yield ()
   }
 
-  private def testIncrementalStep(
-      backward: Boolean,
-      classDefs: Boolean => Seq[ClassDef],
-      moduleInitializers: Boolean => List[ModuleInitializer],
+  def testIncrementalSteps(
+      contextMessage: String,
+      steps: Int,
+      stepToClassDefs: Int => Seq[ClassDef],
+      stepToModuleInitializers: Int => List[ModuleInitializer],
       config: StandardConfig = StandardConfig())(
       implicit ec: ExecutionContext): Future[Unit] = {
 
-    val outputInc = MemOutputDirectory()
-    val outputBatch = MemOutputDirectory()
+    require(steps >= 2, s"require at least 2 steps but got $steps")
+
+    val inputSequence =
+      List.tabulate(steps)(step => (stepToClassDefs(step), stepToModuleInitializers(step)))
+    testIncrementalSequence(contextMessage, inputSequence, config)
```

Comment:
Does it make sense to do that if afterwards you keep a step counter anyways?

Reply:
I'm not sure I understand the question. Do you mean the decomposition between `testIncrementalSteps` and `testIncrementalSequence`? So suggesting to re-inline the behavior of `testIncrementalSequence` into `testIncrementalSteps`?

Reply:
Yeah, that's what I meant. If we'll keep an iteration variable anyways, we might as well use it :)